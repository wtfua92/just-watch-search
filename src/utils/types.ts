export interface SearchItemInterface {
  id: number;
  title: string;
  poster: string;
  year: number;
  type: string;
}

export interface SearchItemRawInterface {
  id: number;
  title: string;
  poster: string;
  original_release_year: number;
  object_type: string;
}

export type SearchResultStateInterface = {
  searchResult: SearchItemInterface[];
  loading: boolean;
};
