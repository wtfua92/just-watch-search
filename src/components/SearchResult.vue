<template>
  <div class="search-result">
    <LoadingSpinner v-if="isLoading" />
    <SearchItem
      v-else
      v-for="item in searchResultItems"
      v-bind="item"
      :key="item.id"
    />
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import SearchItem from "@/components/SearchItem.vue";
import LoadingSpinner from "@/components/LoadingSpinner.vue";
import { mapGetters } from "vuex";

export default Vue.extend({
  name: "SearchResult",
  components: {
    SearchItem,
    LoadingSpinner
  },
  computed: {
    ...mapGetters(["searchResultItems", "isLoading"])
  }
});
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import "src/styles/utils";

.search-result {
  width: 100%;
  display: flex;
  flex-direction: column;
  margin-top: 12rem;

  @include screen-sm {
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-around;
    margin-top: 15rem;
  }
}
</style>
