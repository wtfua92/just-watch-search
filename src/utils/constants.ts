export const BASE_URL = "https://apis.justwatch.com/content/titles/en_US";
// titles
export const IMAGE_DATA = {
  IMAGE_PROFILE: "s332",
  IMAGE_URL: "https://images.justwatch.com"
};
