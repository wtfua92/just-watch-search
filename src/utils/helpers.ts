export const requestCacheWrapper = (requestFunc: Function) => {
  const cache = new Map();

  return (searchValue: string) => {
    if (cache.has(searchValue)) {
      const cachedResult = cache.get(searchValue);

      if (cachedResult.length > 0) {
        return cache.get(searchValue);
      }
    }

    return requestFunc(searchValue)
      .then((result: any) => {
        cache.set(searchValue, result);
        return result;
      })
      .catch((e: Error) => {
        // TODO implement error logging;
        console.log(e);
        return [];
      });
  };
};
