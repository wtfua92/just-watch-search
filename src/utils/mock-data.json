[{"jw_entity_id":"ts53864","id":53864,"title":"Ozark","full_path":"/us/tv-show/ozark","full_paths":{"SHOW_DETAIL_OVERVIEW":"/us/tv-show/ozark"},"poster":"/poster/10929406/{profile}","original_release_year":2017,"tmdb_popularity":12.822258,"object_type":"show","offers":[{"type":"aggregated","monetization_type":"flatrate","provider_id":8,"currency":"USD","urls":{"standard_web":"http://www.netflix.com/title/80117552","deeplink_android_tv":"81037141","deeplink_fire_tv":"intent://www.netflix.com/watch/81037141#Intent;launchFlags=0x00800000;scheme=https;package=com.netflix.ninja;S.source=30;end","deeplink_tvos":"nflx://www.netflix.com/watch/80117552"},"presentation_type":"sd","element_count":3,"new_element_count":3},{"type":"aggregated","monetization_type":"flatrate","provider_id":8,"currency":"USD","urls":{"standard_web":"http://www.netflix.com/title/80117552","deeplink_android_tv":"81037141","deeplink_fire_tv":"intent://www.netflix.com/watch/81037141#Intent;launchFlags=0x00800000;scheme=https;package=com.netflix.ninja;S.source=30;end","deeplink_tvos":"nflx://www.netflix.com/watch/80117552"},"presentation_type":"hd","element_count":3,"new_element_count":3},{"type":"aggregated","monetization_type":"flatrate","provider_id":8,"currency":"USD","urls":{"standard_web":"http://www.netflix.com/title/80117552","deeplink_android_tv":"81037141","deeplink_fire_tv":"intent://www.netflix.com/watch/81037141#Intent;launchFlags=0x00800000;scheme=https;package=com.netflix.ninja;S.source=30;end","deeplink_tvos":"nflx://www.netflix.com/watch/80117552"},"presentation_type":"4k","element_count":3,"new_element_count":3}],"scoring":[{"provider_type":"tmdb:score","value":8},{"provider_type":"imdb:score","value":8.3},{"provider_type":"tmdb:popularity","value":12.822258}]},{"jw_entity_id":"ts20333","id":20333,"title":"Petticoat Junction","full_path":"/us/tv-show/petticoat-junction","full_paths":{"SHOW_DETAIL_OVERVIEW":"/us/tv-show/petticoat-junction"},"poster":"/poster/8559213/{profile}","original_release_year":1963,"tmdb_popularity":24.221,"object_type":"show","offers":[{"type":"aggregated","monetization_type":"flatrate","provider_id":9,"currency":"USD","urls":{"standard_web":"https://www.amazon.com/gp/video/detail/amzn1.dv.gti.3aaa620d-9193-546f-6b19-fee131ea6d44?camp=1789&creativeASIN=B01FURIL8U&ie=UTF8&linkCode=xm2&tag=justwatch09-20","deeplink_android_tv":"intent://watch.amazon.com/watch?asin=B01FURIL8U&time=0&territory=US#Intent;package=com.amazon.amazonvideo.livingroom;scheme=https;end","deeplink_fire_tv":"intent://com.amazon.tv.launcher/detail?provider=aiv&providerId=B01FURIL8U#Intent;package=com.amazon.tv.launcher;scheme=amzn;end","deeplink_tvos":"aiv://aiv/watch?gti=amzn1.dv.gti.3aaa620d-9193-546f-6b19-fee131ea6d44&territory=US&time=auto"},"presentation_type":"sd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"buy","provider_id":10,"retail_price":24.99,"currency":"USD","urls":{"standard_web":"https://www.amazon.com/gp/video/detail/amzn1.dv.gti.3aaa620d-9193-546f-6b19-fee131ea6d44?camp=1789&creativeASIN=B01FURIL8U&ie=UTF8&linkCode=xm2&tag=justwatch09-20","deeplink_android_tv":"intent://watch.amazon.com/watch?asin=B01FURIL8U&time=0&territory=US#Intent;package=com.amazon.amazonvideo.livingroom;scheme=https;end","deeplink_fire_tv":"intent://com.amazon.tv.launcher/detail?provider=aiv&providerId=B01FURIL8U#Intent;package=com.amazon.tv.launcher;scheme=amzn;end","deeplink_tvos":"aiv://aiv/watch?gti=amzn1.dv.gti.3aaa620d-9193-546f-6b19-fee131ea6d44&territory=US&time=auto"},"presentation_type":"sd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"buy","provider_id":10,"retail_price":24.99,"currency":"USD","urls":{"standard_web":"https://www.amazon.com/gp/video/detail/amzn1.dv.gti.3aaa620d-9193-546f-6b19-fee131ea6d44?camp=1789&creativeASIN=B01FURIL8U&ie=UTF8&linkCode=xm2&tag=justwatch09-20","deeplink_android_tv":"intent://watch.amazon.com/watch?asin=B01FURIL8U&time=0&territory=US#Intent;package=com.amazon.amazonvideo.livingroom;scheme=https;end","deeplink_fire_tv":"intent://com.amazon.tv.launcher/detail?provider=aiv&providerId=B01FURIL8U#Intent;package=com.amazon.tv.launcher;scheme=amzn;end","deeplink_tvos":"aiv://aiv/watch?gti=amzn1.dv.gti.3aaa620d-9193-546f-6b19-fee131ea6d44&territory=US&time=auto"},"presentation_type":"hd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"buy","provider_id":3,"retail_price":41.79,"currency":"USD","urls":{"standard_web":"https://play.google.com/store/tv/show?amp=&amp=&cdid=tvseason-9jb2tiMH_EOgt9thqmJ-rg&gdid=tvepisode-ZgIHHZH0-NE&gl=US&hl=en&id=IY5nmi-CT0c","deeplink_android_tv":"IY5nmi-CT0c"},"presentation_type":"sd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"buy","provider_id":3,"retail_price":52.29,"currency":"USD","urls":{"standard_web":"https://play.google.com/store/tv/show?amp=&amp=&cdid=tvseason-9jb2tiMH_EOgt9thqmJ-rg&gdid=tvepisode-ZgIHHZH0-NE&gl=US&hl=en&id=IY5nmi-CT0c","deeplink_android_tv":"IY5nmi-CT0c"},"presentation_type":"hd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"ads","provider_id":73,"currency":"USD","urls":{"standard_web":"http://tubitv.com/tv-shows/7999/?utm_source=justwatch-feed&tracking=justwatch-feed","deeplink_android_tv":"intent://tubitv.com/tv-shows/7999/?link-action=play&tracking=justwatch-feed#Intent;launchFlags=0x00800000;scheme=https;component=com.tubitv/.activities.MainActivity;end","deeplink_fire_tv":"intent://tubitv.com/tv-shows/7999/?link-action=play&tracking=justwatch-feed#Intent;launchFlags=0x00800000;scheme=https;component=com.tubitv/.activities.MainActivity;end","deeplink_tvos":"http://tubitv.com/tv-shows/7999/?tracking=apple-feed&link-action=play&tracking=justwatch-feed&parentSeriesId=446"},"presentation_type":"sd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"ads","provider_id":207,"currency":"USD","urls":{"standard_web":"https://therokuchannel.roku.com/details/s-american_classics.en.ZWMwOTI0ZDEtMDliNi00ZWMwLWJmM2MtMmFkMTU4YjRkNDA2"},"presentation_type":"sd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"ads","provider_id":207,"currency":"USD","urls":{"standard_web":"https://therokuchannel.roku.com/details/s-american_classics.en.ZWMwOTI0ZDEtMDliNi00ZWMwLWJmM2MtMmFkMTU4YjRkNDA2"},"presentation_type":"hd","element_count":1,"new_element_count":1},{"type":"aggregated","monetization_type":"flatrate","provider_id":212,"currency":"USD","urls":{"standard_web":"https://www.hoopladigital.com/title/11684264","deeplink_android_tv":"intent://www.hoopladigital.com/title/11684264#Intent;launchFlags=0x00800000;scheme=https;component=com.hoopladigital.android/com.hoopladigital.android.ui.activity.SplashScreenActivity;end","deeplink_fire_tv":"intent://www.hoopladigital.com/title/11684264#Intent;launchFlags=0x00800000;scheme=https;component=com.hoopladigital.android/com.hoopladigital.android.ui.activity.SplashScreenActivity;end"},"presentation_type":"sd","element_count":1,"new_element_count":1}],"scoring":[{"provider_type":"tmdb:score","value":5.4},{"provider_type":"imdb:score","value":7.1},{"provider_type":"tmdb:popularity","value":24.221}]},{"jw_entity_id":"tm245175","id":245175,"title":"Ozark Sharks","full_path":"/us/movie/ozark-sharks","full_paths":{"MOVIE_DETAIL_OVERVIEW":"/us/movie/ozark-sharks"},"poster":"/poster/12126689/{profile}","original_release_year":2016,"tmdb_popularity":5.133,"object_type":"movie","offers":[{"monetization_type":"flatrate","provider_id":212,"currency":"USD","urls":{"standard_web":"https://www.hoopladigital.com/title/12420868","deeplink_android_tv":"intent://www.hoopladigital.com/title/12420868#Intent;launchFlags=0x00800000;scheme=https;component=com.hoopladigital.android/com.hoopladigital.android.ui.activity.SplashScreenActivity;end","deeplink_fire_tv":"intent://www.hoopladigital.com/title/12420868#Intent;launchFlags=0x00800000;scheme=https;component=com.hoopladigital.android/com.hoopladigital.android.ui.activity.SplashScreenActivity;end"},"presentation_type":"sd"}],"scoring":[{"provider_type":"tmdb:popularity","value":5.133},{"provider_type":"tmdb:id","value":407626},{"provider_type":"tmdb:score","value":4},{"provider_type":"imdb:score","value":3.5}]},{"jw_entity_id":"tm325682","id":325682,"title":"Ozarks Legacy & Legend","full_path":"/us/movie/ozarks-legacy-and-legend","full_paths":{"MOVIE_DETAIL_OVERVIEW":"/us/movie/ozarks-legacy-and-legend"},"poster":"/poster/138463959/{profile}","original_release_year":1995,"tmdb_popularity":1.527,"object_type":"movie","localized_release_date":"1995-04-28","scoring":[{"provider_type":"tmdb:id","value":463755},{"provider_type":"imdb:score","value":6.6},{"provider_type":"tmdb:popularity","value":1.527}],"cinema_release_date":"1995-04-28"},{"jw_entity_id":"tm84103","id":84103,"title":"Bound for Glory","full_path":"/us/movie/bound-for-glory","full_paths":{"MOVIE_DETAIL_OVERVIEW":"/us/movie/bound-for-glory"},"poster":"/poster/140918664/{profile}","original_release_year":1976,"tmdb_popularity":8.792,"object_type":"movie","scoring":[{"provider_type":"imdb:score","value":7.3},{"provider_type":"tmdb:popularity","value":8.792},{"provider_type":"tmdb:score","value":7.2},{"provider_type":"tmdb:id","value":42232}]}]