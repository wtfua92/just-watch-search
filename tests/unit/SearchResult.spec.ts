import { render, screen } from "@testing-library/vue";
import SearchResult from "@/components/SearchResult.vue";
import { SearchItemInterface } from "@/utils/types";

const searchItem: SearchItemInterface = {
  id: 0,
  title: "Ozark",
  poster: "/POSTER.jpg",
  type: "show",
  year: 2017,
  rating: 100.234
};

const searchItemsResult = [
  { ...searchItem, id: 1 },
  { ...searchItem, id: 2 }
];

const renderSearchResult = (
  searchItems: SearchItemInterface[],
  isLoading: boolean
) =>
  render(SearchResult, {
    computed: {
      searchResultItems: () => searchItems,
      isLoading: () => isLoading
    }
  });

describe("SearchResult", () => {
  it("should display all search items", () => {
    const container = renderSearchResult(searchItemsResult, false);
    expect(screen.getAllByText(searchItem.title)).toHaveLength(2);
    expect(container.baseElement).toMatchSnapshot();
  });

  it("should display loading spinner if isLoading=true", () => {
    const container = renderSearchResult(searchItemsResult, true);
    expect(screen.getByText("Loading...")).toBeVisible();
    expect(container.baseElement).toMatchSnapshot();
  });
});
