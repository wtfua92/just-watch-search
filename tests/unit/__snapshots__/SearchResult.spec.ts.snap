// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult should display all search items 1`] = `
<body>
  <div>
    <div
      class="search-result"
    >
      <div
        class="search-result__item"
        id="1"
        style="background-image: url(https://images.justwatch.com/POSTER.jpg);"
      >
        <div
          class="search-result__item__details"
        >
          <div
            class="search-result__item__title"
          >
            <span>
              Ozark
            </span>
             
            <div
              class="search-result__item__badges"
            >
              <span
                class="badge search-result__item__type search-result__item__type--show"
                title="Item type"
              >
                show
              </span>
               
              <span
                class="badge search-result__item__rating"
                title="TMDB Popularity Rating"
              >
                100.2 ★
              </span>
            </div>
          </div>
           
          <p
            class="search-result__item__year"
          >
            2017
          </p>
        </div>
      </div>
      <div
        class="search-result__item"
        id="2"
        style="background-image: url(https://images.justwatch.com/POSTER.jpg);"
      >
        <div
          class="search-result__item__details"
        >
          <div
            class="search-result__item__title"
          >
            <span>
              Ozark
            </span>
             
            <div
              class="search-result__item__badges"
            >
              <span
                class="badge search-result__item__type search-result__item__type--show"
                title="Item type"
              >
                show
              </span>
               
              <span
                class="badge search-result__item__rating"
                title="TMDB Popularity Rating"
              >
                100.2 ★
              </span>
            </div>
          </div>
           
          <p
            class="search-result__item__year"
          >
            2017
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`SearchResult should display loading spinner if isLoading=true 1`] = `
<body>
  <div>
    <div
      class="search-result"
    >
      <div
        class="loader"
      >
        
  Loading...

      </div>
    </div>
  </div>
</body>
`;
